[WIP] use templated isl types

skimo-openhub opened the pull request:
Templated isl types require the user to specify the domain and range
universes of isl objects, allowing the compiler to check whether
it makes sense to combine pairs of objects.
This RFC only converts isPromotableToRegistersBelow and some related
functions to illustrate the effect.
The isPromotableToRegistersBelow was already applying operations
correctly, so the code itself did not require any changes.
However, one variable was reused to store different types
of intermediate result and this one had to be split up into
several variables because they now have different types.

Commit:
order*: use templated isl types

Commit:
extractDomainToIds: use templated isl types

Commit:
use templated isl types in schedule tree

Commit:
activeDomainPoints: cast result to templated isl type

At the same time, turn off automatic type deduction in the callers
of activeDomainPoints.  This allows these callers to gradually
move to templated isl types.
The cast itself is removed after the schedule tree
has been converted to templated isl types.

Commit:
Scop::context(): use templated isl types

Commit:
Scop::dependences: use templated isl types

Commit:
prefixScheduleMupa: use templated isl types

An explicit cast is introduced in prefixScheduleMupa,
which will be removed when infixScheduleMupa is converted
to templated isl types.
Automatic type deduction is temporarily turned off in
MappedScop::detectReductions to avoid having to convert
the entire MappedScop::detectReductions together with prefixScheduleMupa.
It will be reintroduced which will be removed when memberRange is converted
to templated isl types.
Automatic type deduction is also temporarily turned off in
partialScheduleMupa.  It will be reintroduced
after the schedule tree has been converted to templated isl types.

Commit:
makeParameterContext: use templated isl types

Use explicit API calls instead of operators because the use of aff_set
is confusing (at least to the author of this commit).
The conversions to isl::PwAff are needed because, unlike the Python
bindings, the C++ bindings do not have a class hierarchy.
The conversions could also be avoided by introducing
an isl_aff_nonneg_set and an isl_aff_pos_set in isl.

Commit:
Body: use templated isl types

Commit:
Scop::makeContext: use templated isl types

Commit:
MappedScop::findBestSync: use templated isl types

Commit:
halide2isl.cc: extractAccess: automatically deduce type of space variables

This will make it easier to switch to templated isl types.

Commit:
MappedScop::blockMappingSchedule: use templated isl types

Commit:
isPromotableToRegistersBelow: use templated isl types

Commit:
halide2isl.cc: extractReduction: use templated isl types

Commit:
promoteToRegistersBelow: use templated isl types

Commit:
ScheduleTreeMapping::Mapping: use templated isl types

Commit:
reductionMapSchedule: use templated isl types

Commit:
accessSubscriptsAreUnrolledLoops: use templated isl types

Commit:
halide2isl.cc: extractAccess: explicitly cast space to isl::Space<>

This will make it easier to gradually introduce the use of templated
isl types by breaking a cycle.
The explicit cast will be removed again after the core halide2isl
has been converted to templated isl types.

Commit:
MappedScop::findBestSync: temporarily remove type deduction

This allows Scop::dependences to be converted
to templated isl types without having to convert MappedScop::findBestSync
at the same time.  Automatic type deduction will be reintroduced
when MappedScop::findBestSync is converted.

Commit:
partialSchedule: use templated isl types

An explicit cast is introduced in partialSchedule,
which will be removed after the schedule tree
has been converted to templated isl types.
Automatic type deduction is temporarily turned off in
promoteToSharedBelow and promoteToRegistersBelow to avoid having to convert
these entire functions together with partialSchedule.
It will be reintroduced when collectMappingsTo
is converted to templated isl types.

Commit:
partialTargetTiles: use templated isl types

Commit:
core halide2isl: use templated isl types

Commit:
partialScheduleMupa: use templated isl types

An explicit cast is introduced in partialScheduleMupa,
which will be removed after the schedule tree has been converted
to templated isl types.
Automatic type deduction is temporarily turned off in promoteToSharedBelow
to avoid having to convert the entire promoteToSharedBelow
together with partialScheduleMupa.  It will be reintroduced
after the schedule tree has been converted to templated isl types.

Commit:
constructTensorTuple: use templated isl types

Commit:
partialScheduleMupa: use templated isl types

Commit:
Scop::makeContext: explicitly cast space to isl::Space<>

This will make it easier to gradually introduce the use of templated
isl types by breaking a cycle.
The explicit cast will be removed again after Scop::Domain()
has been converted to templated isl types.

Commit:
Scop::makeContext: remove explicit cast

Now that Scop::Domain() has been converted to templated isl types,
the cast is no longer required.

Commit:
accessSubscriptsAreUnrolledLoops: temporarily remove type deduction

This allows TensorReferenceGroup.originalAccesses() to be converted
to templated isl types without having to convert
accessSubscriptsAreUnrolledLoops at the same time.
Automatic type deduction will be reintroduced
when accessSubscriptsAreUnrolledLoops is converted.

Commit:
memberRange: use templated isl types

An explicit cast is introduced in memberRange,
which will be removed when schedule trees are converted
to templated isl types.

Commit:
memory_promotion.cc: tensorElementsSet: explicitly cast space to isl::Space<>

This will make it easier to gradually introduce the use of templated
isl types by breaking a cycle.
The explicit cast will be removed again after Scop::Domain()
has been converted to templated isl types.

Commit:
Scop::makeIslAffFromStmtExpr: use templated isl types

Commit:
activeDomainPointsBelow: cast result to templated isl type

At the same time, turn off automatic type deduction
in the accessSubscriptsAreUnrolledLoops caller
of activeDomainPointsBelow.  This allows this caller to
move to templated isl types separately.
The cast itself is removed after the schedule tree
has been converted to templated isl types.

Commit:
codegen_llvm.cc: getTensorSize: use templated isl types

Commit:
constructThreadToWarp: use templated isl types

Commit:
isSingleReductionWithin: use templated isl types

An explicit cast is introduced in MappedScop::detectReductions,
which will be removed after the schedule tree has been converted
to templated isl types.

Commit:
activeDomainPointsHelper: use templated isl types

Commit:
tensorElementsSet: use templated isl types

Commit:
TensorReferenceGroup::originalAccesses use templated isl types

Commit:
prefixSchedule: use templated isl types

An explicit cast is introduced in prefixSchedule,
which will be removed after the schedule tree
has been converted to templated isl types.
Automatic type deduction is temporarily turned off in
collectDomain and unroll to avoid having to convert
these entire functions together with prefixSchedule.
It will be reintroduced when the schedule tree
is converted to templated isl types.

Commit:
Scop::context: remove explicit cast

Now that halide2isl::makeParamContext has been converted
to templated isl types, the cast is no longer required.

Commit:
Scop::domainRef: use templated isl types

Commit:
separatedOut: use templated isl types

Commit:
canOrder*: use templated isl types

Commit:
labelExtension: use templated isl types

Commit:
labelFilterFromExtension: use templated isl types

Commit:
Scop::domain(): use templated isl types

This commit introduces a cast, which will be removed after
schedule trees have been converted to templated isl types.

Commit:
reductionUpdates: use templated isl types

An explicit cast is introduced in reductionUpdates,
which will be removed when Body is converted
to templated isl types.

Commit:
memory_promotion.cc: tensorElementsSet: remove explicit cast

Now that Scop::Domain() has been converted to templated isl types,
the cast is no longer required.

Commit:
makeSpecializationSet: use templated isl types

Use explicit API calls instead of operators because the use of aff_set
is confusing (at least to the author of this commit).
The conversion to an isl::PwAff is needed because, unlike the Python
bindings, the C++ bindings do not have a class hierarchy.
The conversion could also be avoided by introducing
an isl_aff_zero_set in isl.

Commit:
findCorePositions: use templated isl types

Commit:
MappedScop::threadMappingSchedule: use templated isl types

nicolasvasilache commented:
From a quick glance, this looks great to me

skimo-openhub commented on tc/core/polyhedral/cuda/memory_promotion_heuristic.cc:
```diff
@@ -365,7 +371,7 @@ bool accessSubscriptsAreUnrolledLoops(
     auto unrolledDims = isl::union_pw_aff_list(leaf->ctx_, 1);
-    for (auto node : ancestors) {
+    for (const detail::ScheduleTree* node : ancestors) {
       auto band = node->elemAs<detail::ScheduleTreeElemBand>();
```
Writing this as `auto band = node->template elemAs<detail::ScheduleTreeElemBand>();` allows me to keep the `auto`, but it is again not obvious why I need to introduce `template` here.

ftynse replied:
Syntactic disambiguation.  A particularly nasty interplay between templates and type inference.  The syntax of a template is checked when parsing its definition.  But type inference for `auto` is performed at template instantiation time.  When parsing `elemAs<detail::ScheduleTreeElemBand`, the compiler wouldn't know if  `<` is a comparison operator or the start of a template argument list.  Unless `elemAs` is known to be a template (but it is not known because it's type has not been inferred yet when `node` is declared as `auto`), `<` is treated as a comparison operator.  In this case, the RHS must be an expression. `detail::ScheduleTreeElemBand>()` is not a valid expression so the compiler complains.  Adding the `template` prefix indicates that `<` must be always treated as the start of the template argument list, so it fixes the problem.  So does manually specifying the type of `node` because the compiler now knows that `elemAs` is a function template.

skimo-openhub commented:
On Fri, Jun 22, 2018 at 04:52:38AM -0700, ftynse wrote:
> Syntactic disambiguation.  A particularly nasty interplay between templates and type inference.  The syntax of a template is checked when parsing its definition.  But type inference for `auto` is performed at template instantiation time.  When parsing `elemAs<detail::ScheduleTreeElemBand`, the compiler wouldn't know if  `<` is a comparison operator or the start of a template argument list.  Unless `elemAs` is known to be a template (but it is not known because it's type has not been inferred yet when `node` is declared as `auto`), `<` is treated as a comparison operator.  In this case, the RHS must be an expression. `detail::ScheduleTreeElemBand>()` is not a valid expression so the compiler complains.  Adding the `template` prefix indicates that `<` must be always treated as the start of the template argument list, so it fixes the problem.  So does manually specifying the type of `node` because the compiler now knows that `elemAs` is a function template.

Thanks.  It makes sense now.

skimo

nicolasvasilache commented on tc/core/polyhedral/cuda/memory_promotion_heuristic.cc:
```diff
@@ -365,7 +371,7 @@ bool accessSubscriptsAreUnrolledLoops(
 
     auto unrolledDims = isl::union_pw_aff_list(leaf->ctx_, 1);
-    for (auto node : ancestors) {
+    for (const detail::ScheduleTree* node : ancestors) {
```
try (I think):
```
 auto band = node.template operator->(elemAs<detail::ScheduleTreeElemBand>());
```

Not sure if it works I seem to remember using that with operator.

nicolasvasilache replied:
or what @ftynse explained :)

ftynse commented:
I like the general idea.  What worries me is the need to manually introduce member functions in the class templates.  Those in isl bindings are autogenerated...

skimo-openhub commented:
On Fri, Jun 29, 2018 at 05:50:56AM -0700, ftynse wrote:
> I like the general idea.  What worries me is the need to manually introduce member functions in the class templates.  Those in isl bindings are autogenerated...

The idea would be for us to experiment with this for a while and
if it turns out to be useful, we can think of autogenerating
these functions too.  Note that this would require some
additional annotations.

skimo

Commit:
inline extractDomainToIds

This function will be templated on the mapping type and therefore
needs to move to a header file.
First move it in its original form to make it easier to see
the differences later on.

Commit:
generate templated isl types

Commit:
memory_promotion.cc: tensorElementsSet: temporarily remove type deduction

This allows halide2isl::makeIslAffFromExpr to be converted
to templated isl types without having to convert tensorElementsSet
at the same time.  Automatic type deduction will be reintroduced
when tensorElementsSet is converted.

Commit:
Scop::makeIslAffFromStmtExpr: temporarily remove type deduction

This allows halide2isl::makeIslAffFromExpr to be converted
to templated isl types without having to convert Scop::makeIslAffFromStmtExpr
at the same time.  Automatic type deduction will be reintroduced
when Scop::makeIslAffFromStmtExpr is converted.

Commit:
inline isSingleReductionWithin

This function will be templated on the prefix type and therefore
needs to move to a header file.
First move it in its original form to make it easier to see
the differences later on.

Commit:
inline ScheduleTreeBand::memberRange

This function will be templated on the target type and therefore
needs to move to a header file.
First move it in its original form to make it easier to see
the differences later on.

Commit:
bump isl for export of isl_aff_add_constant_val

Commit:
bump isl for export of isl_map_zip

Commit:
MappedScop::insertMappingContext: drop redundant universe() call

skimo-openhub commented:
On Fri, Jun 29, 2018 at 05:50:56AM -0700, ftynse wrote:
> I like the general idea.  What worries me is the need to manually introduce member functions in the class templates.  Those in isl bindings are autogenerated...

I'm now generating these as well, but I'm still trying
to pick the pieces apart, so it's not quite ready yet.

skimo

Commit:
fix return type of operator+(int i, T A)

This was missing from 9904d3017 (generalize operator+(int i, T A)
to template form, Tue Jul 24 17:45:43 2018 +0200).

Commit:
generalize operator+(T A, int i) to template form

This will ensure that it also applies to templated isl types derived
from isl::aff.
Make sure the template type T has an add_constant_si method to avoid
the template operator getting triggered on vector types.

Commit:
promotionImprovesCoalescing: use get_map_list

This will make it easier to switch to templated types.

Commit:
makeNextElementMap: use templated isl types

Commit:
hasReuseWithin: use templated isl types

Commit:
ScheduleTreeBand::drop: do not call domain() on a set space

The space of an isl_multi_union_pw_aff is the common range space,
i.e., a set space.  The current version of isl tolerates calling
domain() on a set space, but it is technically incorrect.
Call params() instead.
Moving to templated isl types will make this issue more evident.

Commit:
collectMappingsTo: use templated isl types

Commit:
collectMappingsTo: cast result to templated isl type

This allows callers of collectMappingsTo to move to templated isl types.
The case will be removed after the schedule tree
has been converted to templated isl types.

Commit:
accessedWithin: use templated isl types

Commit:
addSingletonReferenceGroups: use get_map_list

This will make it easier to switch to templated types.

Commit:
addSingletonReferenceGroups: use templated isl types

Commit:
addSingletonReferenceGroup: use templated isl types

Commit:
makeSingleton: use templated isl types

Commit:
TensorReference: use templated isl types

Commit:
promotedFootprint: use templated isl types

Commit:
generalize operator/(S left, T right) to template form

This will ensure that it also applies to templated isl types derived
from isl::multi_aff and isl::multi_val.
Make sure the template type T has a scale_down method to avoid
the template operator getting triggered on other types.

Commit:
outputRanges: use templated isl types

Since removeRangeStrides has not been converted yet,
a separate variable is introduced to store its result.
Since ScopedFootprint::strideOffsets is modified,
TensorReferenceGroup::promotion needs some modifications too.

Commit:
removeRangeStrides: use templated isl types

Commit:
TensorReferenceGroup::promotion: temporarily remove type deduction

This allows outputRanges to be converted
to templated isl types without having to convert
TensorReferenceGroup::promotion at the same time.
Automatic type deduction will be reintroduced
when TensorReferenceGroup::promotion is converted.

Commit:
ScopedFootprint.box: use templated isl types

Commit:
TensorReferenceGroup::promotion: use templated isl types

Commit:
Scop::makeScop: automatically deduce type of space variable

This will make it easier to switch to templated isl types.

Commit:
generalize operator-(T A, int i) to template form

This will ensure that it also applies to templated isl types derived
from isl::aff.
Make sure the template type T has an add_constant_si method to avoid
the template operator getting triggered on vector types.

Commit:
generalize operator-(int i, T A) to template form

This will ensure that it also applies to templated isl types derived
from isl::aff.
Make sure the template type T has an add_constant_si method to avoid
the template operator getting triggered on vector types.

Commit:
infixScheduleMupa: use templated isl types

Commit:
inline partialScheduleMupa

This function will be templated on the schedule type and therefore
needs to move to a header file.
First move it in its original form (with an extra "using")
to make it easier to see the differences later on.

Commit:
inline prefixScheduleMupa

This function will be templated on the schedule type and therefore
needs to move to a header file.
First move it in its original form to make it easier to see
the differences later on.

Commit:
inline infixScheduleMupa

This function will be templated on the schedule type and therefore
needs to move to a header file.
First move it in its original form (with an extra "using")
to make it easier to see the differences later on.

Commit:
inline prefixSchedule and partialSchedule

These functions will be templated on the schedule type and therefore
need to move to a header file.
First move them in their original form to make it easier to see
the differences later on.

Commit:
build generate_isl_cpp_h unconditionally

In particular, also build this target when WITH_BINDINGS is not set.
In this case, it currently does nothing and does not depend
on anything, but the availability of an unconditional target
will make it possible to reuse this target to also build
the templated isl header in the next commit, independently
of the WITH_BINDINGS setting.

Commit:
ScheduleTree::makeMappingUnsafe: explicitly cast mapping

This will make it easier to gradually introduce the use of templated
isl types by breaking a cycle.
The explicit cast will be removed again when makeMapping
is converted to templated isl types.

Commit:
makeMapping: use templated isl types

Commit:
MappedScop::map: use templated isl types

Commit:
inline extendSchedule

This function will be templated on the schedule type and therefore
needs to move to a header file.
First move it in its original form (with an extra "using")
to make it easier to see the differences later on.

Commit:
memory_promotion.cc: referenceOriginalAccessesImpl: add missing params() call

In theory, the isl::union_map::empty factory method expects
a parameter space.  In practice, isl acceps any kind of space
(from which it extracts a parameter space).
For switching to templated isl types, it is better
to use a consistent type of space.

Commit:
insertCopiesUnder: use templated isl types

Commit:
generalize operator*(T A, isl::val v) to template form

This will ensure that it also applies to templated isl types derived
from isl::aff.
Make sure the template type T has a scale method to avoid
the template operator getting triggered on other types.

Commit:
TensorReferenceGroup::approximateScopedAccesses: use templated isl types

Commit:
operator+(isl::aff A, isl::val v): call isl::aff::add_constant

There is no need to create an extra isl::aff object here.

Commit:
generalize operator+(T A, isl::val v) to template form

This will ensure that it also applies to templated isl types derived
from isl::aff.
Make sure the template type T has an add_constant method to avoid
the template operator getting triggered on other types.

skimo-openhub commented:
I got kicked out of the GitHub Facebook organization yesterday (before the end of my working day),
so I can't update this PR.  I created a new one instead (#604).

nicolasvasilache reviewed (approved):
trying to unsubscribe, don't see a way other than approving
